### ISSUE TITLE
Use generic java.nio.files interfaces in FileSystemStorage

### ISSUE THREAD

**Issue by ogregoire** [atomashpolskiy/bt#21]
The API as shown now still uses the outdated `java.io.File` API. Since this is a Java-8 only library, it should support the more recent `java.nio` package for better client write-to-disk performances.

**Comment by MorrisLaw** [comment]
Is this enhancement being worked on, or is it up for grabs?

**Comment by magJ** [comment]
Another advantage of using `java.nio.file` is that you can use an alternative `FileSystem` implementation like the in-memory filesystem [jimfs](https://github.com/google/jimfs)

**Comment by atomashpolskiy** [comment]
@MorrisLaw , Working with files is completely isolated inside Storage, so it should not be hard to write an alternative version, making it a good candidate for contributing

**Comment by atomashpolskiy** [comment]
Also see #9 and the last comment about jimfs, might be worth looking into

**Comment by ogregoire** [comment]
@MorrisLaw I wrote this ticket, but didn't start anything yet. So from what I can see, it's indeed up to grabs.

While `java.nio` indeed defines `FileSystem` and friends, the real win is by using `FileChannel` and `MappedByteBuffer`. Especially given how BitTorrent works.

Additionally, the download itself can use java.nio as well to take the full power of multithreading.

**Comment by atomashpolskiy** [comment]
> Additionally, the download itself can use java.nio as well to take the full power of multithreading.

This is a very good point. While network, message encoders/decoders and storage already use NIO channels and buffers, between those the data is transferred as a binary array (encapsulated in `bt.protocol.Piece`). It should very easy to switch to using ByteBuffer instead, so that the data is piped between storage and sockets without creating additional byte[] objects.

**Comment by MorrisLaw** [comment]
@ogregoire @atomashpolskiy Ok guys, I'm going to attempt to tackle this one.

**Comment by MorrisLaw** [comment]
@ogregoire @atomashpolskiy  Just updating everyone on the progress. I'm running into issues with the buildClient(int port) class in examples and tests are failing. Here are some of the changes so far. Any input on the branch I'm working on is also welcome.

https://github.com/MorrisLaw/bt/tree/performance/file-storage

**Comment by atomashpolskiy** [comment]
@MorrisLaw , I've looked through your changes. I appreciate your effort, but there is a couple of issues with your code. I understand that you're a beginner, so it's fine. Let me give you some tips:

1) You can't cast an object reference to some arbitrary type, just to the same type or a supertype (plus a few more specific cases), otherwise you'll get a `ClassCastException` in runtime. E.g.
```java
// bt/cli/CliClient.java:119
Files.getFileStore((Path) options); // => java.lang.ClassCastException: bt.cli.Options cannot be cast to java.nio.file.Path
```
You might want to skim through the corresponding [JLS Chapter 5. Conversions and Promotions](https://docs.oracle.com/javase/specs/jls/se7/html/jls-5.html), specifically section _5.1.6. Narrowing Reference Conversion_.

**Keep in mind that in most cases casting is bad and probably means that you're doing something wrong.**

2) Some nullity checks in `FileSystemStorageUnit` don't make sense, e.g.
```java
// bt/data/file/FileSystemStorageUnit.java:53
if (Files.createFile(file) == null) { // => always false
  ...
}
```
If you look inside these methods, you'll see that they always return the first argument (`Path` to file) or throw an exception.

3) Another issue with casting:
```java
// bt/data/file/FileSystemStorageUnit.java:185
public long size() {
   ...
   return (long)Files.size(file); // => casting is redundant, Files.size(Path) already returns `long`
   ...
}
```

4) As a general rule you shouldn't (and per this issue don't need to) change public API. Some client code may depend on certain methods/constructors having specific signatures, and changing the signature means breaking the client code, e.g.
```java
    /*
     * @since 1.0
     */
   // bt/data/file/FileSystemStorage.java:62
    public FileSystemStorage(File rootDirectory) { // => can't just change the parameter type to java.nio.file.Path
        ...
    }
```
This one is especially bad:
```java
// bt/data/StorageUnit.java:42
ByteBuffer readBlock(long offset, int length); // => wrapping a byte array into buffer does not make much sense
```
Of course the API is not limited to methods only; it includes class and package names, thrown exceptions, behavior that is promised in the javadoc, even side effects. So in *most* cases it's preferable to only *add* new API without changing/removing the old one (if old API is obsolete, it should be marked as deprecated and noticed in upgrade notes, so that the client code can migrate safely before it's finally removed).

5) `dependency-reduced-pom.xml` is generated by `maven-shade-plugin` and should not be checked into git repository. It's my fault that I haven't added the corresponding rule to `.gitignore`, I've updated it in [add949a](https://github.com/atomashpolskiy/bt/commit/add949a52c5b6f4f16ce18f5d5d2fa957fdc27fb).

**To sum up**: I suggest that you discard your changes and start over, rebasing onto the latest commit in master. Keeping the above comments in mind, you need to do just the following:
- add a new constructor `FileSystemStorage(Path)`. Old constructor `FileSystemStorage(File)` should delegate object creation to the new constructor by calling `this(rootDirectory.toPath())`
-  update `FileSystemStorageUnit` internals to use `java.nio.files` interfaces
- verify that all tests pass

**Comment by MorrisLaw** [comment]
Thank you for the thorough response !

**Comment by atomashpolskiy** [comment]
By the way, it turns out `FileSystemStorageUnit` already uses `java.io.RandomAccessFile` and `java.nio.channels.FileChannel`, so the performance is not a problem. What's this issue is really about is switching to using generic `java.nio.files.Path` interface instead of `java.io.File`, which will let users use `FileSystemStorage` as a traditional file storage or an in-memory storage (e.g. via [jims](https://github.com/google/jimfs)), depending on their needs. We might also want to switch to using jimfs-backed version in `bt-tests` to speed up things and remove the need to cleanup the test files after each test (as a bonus this will probably fix the Windows build).

**Comment by ogregoire** [comment]
There is no need to use `RandomAccessFile`. If one uses nio everywhere but a RAF when it's written on the disk, all the performance gain is lost...

Here's how to work with RAF using nio:

    FileChannel fileChannel = FileChannel.open(file, READ, WRITE);

    ByteBuffer buffer = ...
    fileChannel.position(12345); // start writing at position 12345
    while (buffer.hasRemaining()) {
      fileChannel.write(buffer);
    }

You can get a `Path` from a `File` using `File::toPath`.

Finally regarding the interface(s), indeed, you need to keep the existing, but you can create another method next to it with a `default` behavior:

    interface Abc {
      void doSomething(File f);
    }

Will become:

    interface Abc {
      @Deprecated default void doSomething(File f) {
        doSomething(f.toPath());
      }
      void doSomething(Path p);
    }

Then, of course, you need to rewrite all methods that use `Abc`, but that's part of the job for this ticket.

If needed, I can work on this issue, but not before next week.

**Comment by atomashpolskiy** [comment]
> There is no need to use RandomAccessFile. If one uses nio everywhere but a RAF when it's written on the disk, all the performance gain is lost...

We do want to get rid of `RandomAccessFile` for API reasons, but don't expect any performance gains, because it's already used in combination with `FileChannel`, like this:
```java
raf.seek(offset);
raf.getChannel().write(buffer);
```

> Finally regarding the interface(s), indeed, you need to keep the existing, but you can create another method next to it with a default behavior...

Valid point, but we were discussing constructors ¯\_(ツ)_/¯

**Comment by ogregoire** [comment]
Yeah, but I was mistaken (it's been a long time I worked with io in general), there is no need for a `FileChannel`, but for a `MappedByteChannel` (I edited my previous comment). And for that, it's rather bad to always create such a channel: since the channel will be loaded from disk everytime and therefore the performance lost. Keep only a `MappedByteChannel`, which will load the appropriate region upon need properly all by itself.

**Comment by atomashpolskiy** [comment]
@ogregoire , makes sense now, but this behavior is not always beneficial, e.g. when streaming media I want the data to be written to disk immediately. So it should be explicitly configurable by the user (either by some flag or better by instantiating a different implementation class). Common parts can be extracted into an abstract supertype

**Comment by ogregoire** [comment]
You can simply use `force()` after each "write" command. That will flush the changes to the disk, but keep that region in memory. Really, it's that simple.

**Comment by atomashpolskiy** [comment]
I get that but my point was that whether flush is eventual or immediate should be conditional and based on user's preference and/or current mode of operation.

There is also a problem of how much memory can/should be used. We can't just map all data into memory, so there should be some centralized management of what is loaded and what is discarded at each given moment. 
Finally, this is also a design issue because StorageUnit is supposed to be just that - a storage, not a flushable cache.

Initially I was thinking about buffering data at the I/O worker level, which has (indirect) access to the storage and can decorate it as needed. Additionally it has the statistics of when/how often each piece is requested, so it can act like a full-fledged cache, re-using the same mapped buffers.

**Comment by ogregoire** [comment]
I guess there must be some misunderstanding here: it's entirely possible to have a storage that works as you expect using nio and channels and mappedbytebuffer. The thing is that currently you always go through a byte[], which is far from efficient. The idea is to get rid of this temporary buffer by using the buffers we get.

Regarding memory, it's a mappedbytebuffer, so it's memory that's directly mapped to the drive. No system calls. I don't see why it should be a user-preference: the user will want it to be written to disk, according to the best way to handle the disk. This is what is done by default.

Know what? I'll fork, put my changes in, and show them so if you like that, let's merge, if not, maybe use it as a base? But as I said, I can't start before Monday at the earliest. Oh, and I can't guarantee I'll be able to answer anymore today or tomorrow.

**Comment by MorrisLaw** [comment]
@ogregoire If you already have a solution maybe you can complete the ticket Monday. Otherwise as far as @atomashpolskiy original suggestion, itseemed like something I can achieve sometime this weekend. But reading through these comments, it looks like it's going to be more involved.

Is this still the case:
`What this issue is really about is switching to using generic java.nio.files.Path interface instead of java.io.File` 

Or is there a different issue or just different opinions on how to fix the same issue ?

**Comment by atomashpolskiy** [comment]
@ogregoire ,
> I guess there must be some misunderstanding here...

There might be a misunderstanding, because

> The thing is that currently you always go through a byte[], which is far from efficient.

I'm not sure where this comes from; there's a `FileChannel`, that is passed a `ByteBuffer`. No byte arrays are involved at this level.

> The idea is to get rid of this temporary buffer by using the buffers we get from the MappedByteBuffer.

> Regarding memory, it's a mappedbytebuffer, so it's memory that's directly mapped to the drive. No system calls. I don't see why it should be a user-preference: the user will want it to be written to disk, according to the best way to handle the disk. This is what is done by default.

I completely understand your idea. I'm just trying to argue, that:

1) We don't **always** want to create a `MappedByteBuffer`. Specifically we don't want to map anything to memory when we're streaming (what's the point of having an intermediate in-memory buffer, if we want data to be flushed to disk immediately and call `force()` after each `write()`? Seems like wasting RAM to me, directly working with `FileChannel` is clearly a superior option in this case)

2) When we do want to have an intermediate buffer in the form of `MappedByteBuffer` (e.g. when we're downloading a 100GB collection of files and don't care, if the data is on disk on in memory, until the download is finished), we also want to control and limit **the total amount of RAM being used**. Simply mapping each individual file to memory isn't going to work, which means that creating a `MappedByteBuffer` **inside** the storage unit (which is equivalent to a file in Bt terms) does not make much sense. It should be done in some upper layer, that manages all read/write operations.

I might be communicating not clearly enough though, excuse me if so, because English is not my first language :)

---

@MorrisLaw ,
> Is this still the case:
What this issue is really about is switching to using generic java.nio.files.Path interface instead of java.io.File

Absolutely. Whatever we do with regards to intermediate buffering will be built upon the initial refactoring of `FileSystemStorageUnit` to get rid of `java.io.File` (though there will still be a constructor `FileSystemStorage(File)` for backwards compatibility reasons).

**Comment by atomashpolskiy** [comment]
Wow, this whole topic is a hell of a lot of words. I personally am going to go and read something about mapped memory, because it might be that I just misunderstand the concept and how it works. Cheers

**Comment by MorrisLaw** [comment]
Here's another crack at the enhancement. Is this acceptable ?
https://github.com/MorrisLaw/bt/tree/enhancement/storage-nio

**Comment by MorrisLaw** [comment]
Actually some tests failed, so I'm looking into that now. I only tested bt-core at first, then when running all of them it failed the build.

```
Tests run: 249, Failures: 0, Errors: 2, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Bt ................................................. SUCCESS [  0.024 s]
[INFO] Bt BEncoding ....................................... SUCCESS [  1.427 s]
[INFO] Bt Core ............................................ SUCCESS [  1.115 s]
[INFO] Bt HTTP Tracker Client ............................. SUCCESS [  0.447 s]
[INFO] mldht - Bittorrent Mainline DHT library ............ SUCCESS [  2.690 s]
[INFO] Bt Tests ........................................... FAILURE [  1.932 s]
[INFO] Bt DHT ............................................. SKIPPED
[INFO] Bt libmldht adapter ................................ SKIPPED
[INFO] Bt CLI Launcher .................................... SKIPPED
[INFO] Examples ........................................... SKIPPED
[INFO] JaCoCo aggregate report ............................ SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 8.377 s
[INFO] Finished at: 2017-08-06T17:44:54-04:00
[INFO] Final Memory: 26M/276M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.18:test (default-test) on project bt-tests: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/morrislaw/Projects/bt/bt-tests/target/surefire-reports for the individual test results.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :bt-tests
```

**Comment by MorrisLaw** [comment]
Ahaha nvm, it looks like I ran it in the directory without the pom file. Now it all passes.
```
Tests run: 249, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 4.016 s
[INFO] Finished at: 2017-08-06T17:50:55-04:00
[INFO] Final Memory: 14M/250M
[INFO] ------------------------------------------------------------------------
```

**Comment by MorrisLaw** [comment]
Still working on errors, it appears that I'm failing the CI tests.

**Comment by ogregoire** [comment]
@MorrisLaw Just a small note from a very superficial review: you used `Files.createDirectory(parent)` where you should actually use `Files.createDirectories(parent)`. The difference between the two is that  if you call the first method with `a/b/c` when only `a` exists, you'll get an error because `b` doesn't exist, where with the second method, you'll get `c` and all its missing parents created without troubles.

**Comment by MorrisLaw** [comment]
Thank you for spotting that. I will refactor the code.

**Comment by ogregoire** [comment]
Okay, I reviewed the code, and added several comments ([see list with a comment icon](https://github.com/atomashpolskiy/bt/compare/master...MorrisLaw:enhancement/storage-nio)).

I still do think it's better to use `MappedByteBuffer` for performance, (compared to a simple, unmapped `FileChannel`). But to be honest, that would require some refactoring in the `StorageUnit` interface to use it optimally. With the interface as is, what you wrote, @MorrisLaw  is the best we can do.

**Comment by atomashpolskiy** [comment]
@ogregoire , I'm genuinely interested, what you have in mind, and I'm all for experimentation, as long as it's not the default behavior.

**Comment by ogregoire** [comment]
Well, basically there's currently a design flaw. It's a subtle one, so what you did was actually good, but in case of using `ByteBuffer`s, it's a bit of a premature optimization.

So I would suggest using such an interface (comments removed for brievity):

	public interface StorageUnit extends AutoCloseable {

		ByteBuffer readBlock(long offset, long size);

		void writeBlock(ByteBuffer buffer, long offset);

		long capacity();

		long size();
	}

Obviously I deleted the `byte[]`. Why is that? Because it's trivial to wrap a `byte[]` into a `ByteBuffer` and it's also trivial to get a `byte[]` from a `ByteBuffer`.

Okay, now why in the world would I make the read operation return a new ByteBuffer everytime? Well, this is where the flaw is in the original design: it was assumed that creating your own ByteBuffer and reusing it again and again will provide you with improved performance regarding memory. It makes a lot of sense... Until we know how `ByteBuffer` works internally, and more specifically the `MappedByteBuffer` returned by `FileChannel.map`.

The important part of [`ByteBuffer`'s documentation](http://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html) is the following (slightly edited, emphasis mine):

> ## Direct vs. non-direct buffers 
> A byte buffer is either direct or non-direct. Given a direct byte buffer, the Java virtual machine will make a best effort to perform native I/O operations directly upon it. That is, it **will attempt to avoid copying the buffer's content** to (or from) an intermediate buffer before (or after) each invocation of one of the underlying operating system's native I/O operations. 
>  ...
> **A direct byte buffer may also be created by mapping a region of a file directly into memory**. An implementation of the Java platform may optionally support the creation of direct byte buffers from native code via JNI. If an instance of one of these kinds of buffers refers to an inaccessible region of memory then an attempt to access that region will not change the buffer's content and will cause an unspecified exception to be thrown either at the time of the access or at some later time. 

This means that if we use the byte buffers that we get from `FileChannel.map`, the performance can be boosted. Especially when dealing with large, long-open files. Oh, by the way, what is mostly shared through torrents? Yes: large files. And they're open for a long time when they're being downloaded or uploaded.

Concretely, this means that we will provide directly the mapped file-region to the network buffer. Meaning that we skip Java's heap entirely by not copying `byte[]` into other temporary `byte[]` more than absolutely necessary.

**Comment by atomashpolskiy** [comment]
Ok, we definitely need more details here. As already mentioned in this topic, [FileChannel](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/sun/nio/ch/FileChannelImpl.java) class has a [map()](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/sun/nio/ch/FileChannelImpl.java#l808) method, that is full of black magic and trickery (e.g. trying to gracefully handle OutOfMemory errors by asking JVM to perform garbage collection to trigger deallocation of previously allocated native memory buffers... you get the idea), but really boils down to a [native method](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/solaris/native/sun/nio/ch/FileChannelImpl.c#l74) and [mmap](http://man7.org/linux/man-pages/man2/mmap.2.html) system call.

What `mmap` does is create a file mapping in the calling process's address space, more specifically in the native memory of the process (off-Java-heap). Upon creation the entire mapping is initialized by kernel with page faults, each of which will be resolved upon first access (i.e. lazy-loading data from disk).

The `FileChannel.map()` method returns a `MappedByteBuffer`(http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/nio/MappedByteBuffer.java). It's not a real buffer per se, but rather an abstract super-type for native memory (off-Java-heap) buffers. It provides a few utility methods, most importantly `isLoaded()`, `load()` and `force()`, all of which are backed by corresponding native methods and boil down to [mincore](http://man7.org/linux/man-pages/man2/mincore.2.html), [madvise](http://man7.org/linux/man-pages/man2/madvise.2.html) and [msync](http://man7.org/linux/man-pages/man2/msync.2.html) system calls.

The actual type of the returned object is `DirectByteBuffer`, which is a flavor of [Direct-X-Buffer](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/nio/Direct-X-Buffer.java.template). This is the only implementation of a native memory buffer (comes in different flavors, depending on the type of content, hence X in template name). It can be constructed in two ways: (1) either by requesting a buffer of a given capacity (e.g. by calling `ByteBuffer#allocateDirect`), which ends up in [malloc](http://man7.org/linux/man-pages/man3/free.3.html), (2) or by providing an existing "buffer" in the form of a virtual address of a pre-allocated memory region.

Considering Java, there are two peculiarities to keep in mind when working with `DirectByteBuffer`:
1) It implements the `Buffer` type, which means its' capacity is restricted to `Integer.MAX_VALUE` (2^31, or ~2GB)
2) Allocated native memory can not be freed deterministically (i.e. on-demand), but only as part of the GC. With long-living large-sized buffer objects this can become a problem, because they are likely to be promoted to the parts of the heap, that are less frequently processed by GC.

Here are the implications:
1) Creating a file mapping for a file of size N requires equal amount of RAM. If there's less physical memory than N, then the kernel will manage what parts of the file are resident in memory at each given moment (based on what pages are accessed), performing page-ins for requested pages and discarding unused pages. I think it's safe to say that memory mapping makes more sense when the whole file fits into memory.
2) Given that, in general case, we don't know beforehand, what is the total size of torrent's data, it's not reasonable to map all data into memory. This will clearly cause OOM at some point.
3) While memory-mapping a file definitely makes sense when we **already have the file** (i.e. when seeding), **it in fact can negatively affect the performance when we're downloading the file**. This is because each write _may_ trigger a page-in from disk, even if there is no data yet (precise ratio of writes to page-ins depends on the size of the block being written, the size of memory and disk pages and whether the corresponding pages have already been loaded).

Therefore, while it may be beneficial to naively memory-map all files in _some cases_, more specifically:
- when total size of files is not very large (which size is "large" is specific to every given environment);
- when there are no particularly large files (> 2GB);
- when we already have the data and are just seeding;

I don't think it's as simple as you want it to be in general case.

**Comment by atomashpolskiy** [comment]
With regards to this:
> Concretely, this means that we will provide directly the mapped file-region to the network buffer. Meaning that we skip Java's heap entirely by not copying byte[] into other temporary byte[] more than absolutely necessary.

You are being overly optimistic about the code :) There are several layers between storage and network, and some of them currently rely on passing each block as `byte[]`.

**Comment by ogregoire** [comment]
> There are several layers between storage and network, and some of them currently rely on passing each block as `byte[]`.

I'm aware of that and I'm not saying the change is trivial or even simple. What I highlighted is only the root for all work.

You have several concerns and I'm aware of them. I was aware of them before you raised them because I had them in the past. as well 

You're saying we're providing `byte[]` at various places in the code. I saw that. The `Range` interface is one example and I see it's used for different purposes, one of them is computing the hash. In that particular case, why is it transmitting a `byte[]` and not a `ByteBuffer`? Because the `ByteBuffer` must be reused? Well, I never see any use of the `mark()` and `reset()` methods which exist for that kind of use-cases, for instance.

Switching to nio is no trivial case and must be done application-wide. There are a lot of changes and perhaps it might be necessary to rethink some ways the library works at its core.

It might be good for a further major version, and not for minor one.

**Comment by atomashpolskiy** [comment]
> why is it transmitting a byte[] and not a ByteBuffer

It won't be a problem to add methods that work with buffers rather than byte arrays. I think it was kind of an unconscious decision to use `byte[]`, because messaging uses `byte[]` to represent a BitTorrent's `piece`. So, the problem is mostly in the messaging layer.

> You're saying **we're** providing byte[] at various places in the code.

I like your attitude! No joking :)

> Switching to nio is no trivial case and must be done application-wide. There are a lot of changes and perhaps it might be necessary to rethink some ways the library works at its core.

As I've said previously, networking already uses NIO... The problem is that working with storage (reads and writes) is asynchronous in respect to networking. In short, current design is based on the following:
1) Messaging is single-threaded; consuming and producing messages from/to peers is done in a single thread in [MessageDispatcher](https://github.com/atomashpolskiy/bt/blob/master/bt-core/src/main/java/bt/net/MessageDispatcher.java)
2) [Storage I/O](https://github.com/atomashpolskiy/bt/blob/master/bt-core/src/main/java/bt/torrent/data/DefaultDataWorker.java) is async and can be parallelized; it uses a single-threaded executor for now, but I think it's safe to use multiple threads here

I agree that piping data between `network <----> storage` is non-trivial in such design. But it also has its' pros.

**Comment by ogregoire** [comment]
> > You're saying we're providing byte[] at various places in the code.
> I like your attitude! No joking :)

Sorry, that's a deformation from French in which "we" can be both "us" and the undefined pronoun (for French-speaking guys, I'm speaking about "on", not "nous").

Regarding the rest, I discovered your repo through your HN post last week, I was impressed and wanted to contribute. I haven't had the time to understand all the design yet. I know how the BitTorrent protocol works in the big lines (b-encoding, distribution of chunks, etc). But I most certainly haven't delved in it like you did to develop this project. However I have some experience with nio, though it's not natural for me, I still prefer good ol' Input/OutputStream.

So indeed I might be naive on some topics, but I hope to help in the best of my knowledge.

**Comment by atomashpolskiy** [comment]
@ogregoire , man, come on, I'm glad that you're contributing, and I'm very thankful. I immensely enjoy our conversation and under no circumstances would like to seem "smarter" or whatever. In fact I'm sure as hell that you are much more knowledgeable than me, and I hope you will continue to share your knowledge. I just wanted this conversation to pave a road from A (where we are) to B (where we'd like to be), so I thought it would be beneficial to start a more in-depth discussion. And I totally imply that I might be wrong in what I'm saying, so you're welcome to correct and provide your own arguments! Peace!

**Comment by ogregoire** [comment]
Hmm, I guess we weren't on the same page: I'm still in ;)

I'll try to work on a prototype where I introduce those changes gradually, if that's okay. Probably in my own branch or something for quite a while. We'll see.

**Comment by atomashpolskiy** [comment]
Sure! I also have some other global tasks in mind, which would require research and quite possibly re-thinking approach and APIs. I think you could do a great job on these, so if you're interested, I'll try to dump my thoughts in a couple of tickets
